source/iterator/iterator.go

```diff
@@ -104,30 +105,75 @@ type Iterator struct {
 
 // Params represents an incoming iterator params for the New function.
 type Params struct {
-	Position       *Position
-	URL            string
-	Table          string
+	Position *Position
+	URL      string
+	Table    string
+
+	SnapshotTable string
+	TrackingTable string
+	Trigger       string
+
 	OrderingColumn string
 	KeyColumns     []string
 	Snapshot       bool
 	Columns        []string
-	BatchSize      int
+
+	BatchSize int
+}
+
+func NewParams(pos *Position, config config.Source) *Params {
+	p := &Params{
+		Position:       pos,
+		URL:            config.URL,
+		Table:          config.Table,
+		OrderingColumn: config.OrderingColumn,
+		KeyColumns:     config.KeyColumns,
+		Snapshot:       config.Snapshot,
+		Columns:        config.Columns,
+		BatchSize:      config.BatchSize,
+	}
+	p.setHelperObjects(config.TrackingPrefix)
+
+	return p
+}
+
+// setHelperObjects sets the names of the helper tables and triggers we use:
+// the snapshot table, the tracking table and the trigger name.
+func (p *Params) setHelperObjects(prefix string) {
+	if p.Position != nil {
+		p.SnapshotTable = p.Position.SnapshotTable
+		p.TrackingTable = p.Position.TrackingTable
+		p.Trigger = p.Position.Trigger
```

Comment:
> So if I have an existing pipeline, I change the tracking prefix and then restart the pipeline, it will have no effect? confused

Good point... IMHO, updating the configurations should be a separate task, because there's a few cases we need to cover there. It's likely outside the scope of this PR. E.g. when you change the prefix, there still might be data in the old tracking table, so we need to figure out what to do with it. On the other hand, if we're not reading the tracking table name from the position, we would read it from the config. We can have a case where the config changed from a custom prefix to an auto-generated one. Currently, the connector README points out that manual changes to the tracking tables can cause unexpected behavior, so unless you disagree it looks like it's better to tackle this separately.





> And I have no control over the random postfix added to the table, so if I somehow lose my Conduit state I have no way to force Conduit to use an existing tracking table that is already full of data.

Unless I'm missing something, even in that case it wouldn't be possible to resume from exactly the same place where the pipeline stopped. We need to know where exactly in the snapshot or the tracking table did the connector stop. That is in the position, which in this case would be lost with the rest of the Conduit state.

Reply:
> Then changing the prefix would have an effect and the user could be in full control over the helping object names.

For the record, we discussed this topic of configuration at a team sync. We want to make the above-mentioned full control over helper object names possible. We also want to make it possible to specify only the prefix and let the connector handle the uniqueness by adding a random (requirement of our internal platform). That means we'll have 3 options:
1. Names completely generated by the connector, i.e. no config provided by a user (e.g. conduit_snapshot_123)
2. Prefix configured by a user (e.g. `prefix=foobar` would give `snapshot_table=foobar_snapshot_456` and `tracking_table=foobar_tracking_456`).
3. Full control over all names (e.g. `snapshot_table=table_s`, `tracking_table=table_t`)

Reply:
> when you change the prefix, there still might be data in the old tracking table, so we need to figure out what to do with it

This will be easier to implement once we have connector lifecycle events. The connector will know when the config changes and react accordingly. If the connector "owns" the helper objects I'd expect it to rename them (or delete and recreate) and ensure that it doesn't lose existing data in the tracking table. If the objects were created by someone else, I'd expect the connector to simply switch to the new ones, regardless what was there before. Either way, I don't think this is strictly required in this issue, but I would like us to consider the bigger picture to make sure the configuration we are adding is what we want.

That reminds me of another thing - the connector currently always creates the table and trigger, but what if it doesn't have permission to do this? We should probably allow the user to create the objects and just use them if they already exist.

> even in that case it wouldn't be possible to resume from exactly the same place where the pipeline stopped.

Yes, we wouldn't necessarily start at the _exact_ same row, we might create some duplicated records. Looking at the code, the connector deletes processed rows from the tracking table every 5 seconds, so in the worst case we would reprocess 5 seconds worth of rows. I'd expect that in a graceful stop the connector removes all successfully processed rows from the tracking table, so unless the pipeline was stopped forcefully, we should still be able to resume exactly where we left off.¹

From how I understand the connector, Oracle is pushing data to the tracking table via the trigger and the connector is just consuming whatever gets written to the table. Once a record is processed, the corresponding row is eventually removed. This seems like the position stored in Conduit doesn't play much of a role, the connector is tracking its own progress in an Oracle table. So now I'm asking myself why do we even need to read anything from the record position? It seems like the connector should simply read the rows from the tracking table and ensure they are deleted when records are acked, then it will always resume at the correct position.

**TL;DR:** I have a couple of concerns about the current behavior of the connector which are clearly out of scope of this issue. At the same time, I'd like to avoid making breaking changes, so I think we should take a moment to define the end goal regarding configuration options, the behavior we expect from the connector for these options, and how the connector behaves if they change. Then we can be sure that the options and behavior we are introducing here are here to stay.

---

¹ Sidenote: I don't think that's the case right now. The last cleanup is triggered in [`Teardown`](https://github.com/conduitio-labs/conduit-connector-oracle/blob/a42b9bc1add95acde578cce609d06c52e8287a74/source/source.go#L157), which sends a [signal to the cleanup routine](https://github.com/conduitio-labs/conduit-connector-oracle/blob/a42b9bc1add95acde578cce609d06c52e8287a74/source/iterator/cdc.go#L228), but the `ctx` used in the [cleanup routine](https://github.com/conduitio-labs/conduit-connector-oracle/blob/a42b9bc1add95acde578cce609d06c52e8287a74/source/iterator/cdc.go#L324) is coming from `Open`, and that context is already closed at the time `Teardown` is called.